### PR TITLE
Add a head.additional block to adminhtml layout

### LINKED ISSUE #29165: [Issue] Add a head.additional block to adminhtml layout
This issue is automatically created based on existing pull request: magento/magento2#28389: Add a head.additional block to adminhtml layout

---------
Module developers are currently able to add external JS files to the admin area's `<head>` section, but adding inline scripts is not as straightforward as it is in the Frontend area because the admin layout doesn't define the `head.additional` block. In fact, I couldn't find a way to add inline scripts to the head in a way that's not "hackish" and remains compatible with other modules. The mechanism to do this easily is simply missing.

This PR addresses the mechanism issue: once this block is added `\Magento\Framework\View\Result\Page` will make sure all `head.additional` child blocks are rendered, mirroring the functionality in the Frontend area.

### Preconditions: (*)
This PR simply adds a `head.additional` block of type `\Magento\Framework\View\Element\Text\ListText` to the Adminhtml's `default` layout (via `magento2/module-base`).

### Related Pull Requests
None

### Fixed Issues (if relevant)
I couldn't find any.

### Steps to reproduce (*)

1. Create a module that adds the following layout update in the **_adminhtml_** area:

```xml
<?xml version="1.0"?>
<page xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
      xsi:noNamespaceSchemaLocation="urn:magento:framework:View/Layout/etc/page_configuration.xsd">
    <referenceBlock name="head.additional">
        <block class="Magento\Framework\View\Element\Template" name="test"
               template="Magento_Backend::page/copyright.phtml"/>
    </referenceBlock>
</page>
```

2. Run: `php bin/magento cache:clean`;
3. Go to Admin and view page source;

**Actual Result:** :heavy_multiplication_x: no copyright text in the `<head>`

Screenshot from 2020-07-16 11-19-05

**Expected Result:** :heavy_check_mark: The copyright text rendered within any admin page's <head> area

#### Why a `ListText` block?
1. Because a `container` renders its children within a tag, and we don't want that inside `head`.
2. Because in the Frontend area this is done with a `Template` block with a `container.phtml` template is used, but that template doesn't exist in the adminhtml area.
3. Because `ListText` simply gets the job done exactly as it's supposed to and without any side-effects.

### Backwards Compatibility
This change adds a new block to an existing layout and therefore shouldn't break backwards compatibility with any existing code.

### Automated tests?
I'd be happy to create them if you really think it's necessary but I'll need a bit of guidance. However, I also doubt this is the kind of things we're testing, because existing tests for `\Magento\Framework\View\Result\Page` and `\Magento\Framework\View\Result\Page` would essentially cover most (if not all) requirements for this functionality.

### Contribution checklist (*)
 - [ ] Pull request has a meaningful description of its purpose
 - [ ] All commits are accompanied by meaningful commit messages
 - [ ] All new or changed code is covered with unit/integration tests (if applicable)
 - [ ] All automated tests passed successfully (all builds are green)

### PR DESCRIPTION
**Pull request by gsomoza**
Module developers are currently able to add external JS files to the admin area's `<head>` section, but adding inline scripts is not as straightforward as it is in the Frontend area because the admin layout doesn't define the `head.additional` block. In fact, I couldn't find a way to add inline scripts to the head in a way that's not "hackish" and remains compatible with other modules. The mechanism to do this easily is simply missing.

This PR addresses the mechanism issue: once this block is added `\Magento\Framework\View\Result\Page` will make sure all `head.additional` child blocks are rendered, mirroring the functionality in the Frontend area.

### Description (*)
This PR simply adds a `head.additional` block of type `\Magento\Framework\View\Element\Text\ListText` to the Adminhtml's `default` layout (via `magento2/module-base`).

### Related Pull Requests
None

### Fixed Issues (if relevant)
I couldn't find any.

### Manual testing scenarios (*)
1. Create a module that adds the following layout update in the  `adminhtml` area:
```
<?xml version="1.0"?>
<page xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
      xsi:noNamespaceSchemaLocation="urn:magento:framework:View/Layout/etc/page_configuration.xsd">
    <referenceBlock name="head.additional">
        <block class="Magento\Framework\View\Element\Template" name="test"
               template="Magento_Theme::html/copyright.phtml"/>
    </referenceBlock>
</page>
```
2. You should see the copyright text rendered within any admin page's `<head>` area.

(which is invalid HTML but should work as a quick test)

#### Why a `ListText` block?
1. Because a `container` renders its children within a tag, and we don't want that inside `head`.
2. Because in the Frontend area this is done with a `Template` block with a `container.phtml` template is used, but that template doesn't exist in the adminhtml area.
3. Because `ListText` simply gets the job done exactly as it's supposed to and without any side-effects.

### Backwards Compatibility
This change adds a new block to an existing layout and therefore shouldn't break backwards compatibility with any existing code.

### Automated tests?
I'd be happy to create them if you really think it's necessary but I'll need a bit of guidance. However, I also doubt this is the kind of things we're testing, because existing tests for `\Magento\Framework\View\Result\Page` and `\Magento\Framework\View\Result\Page` would essentially cover most (if not all) requirements for this functionality.

### Contribution checklist (*)
 - [ ] Pull request has a meaningful description of its purpose
 - [ ] All commits are accompanied by meaningful commit messages
 - [ ] All new or changed code is covered with unit/integration tests (if applicable)
 - [ ] All automated tests passed successfully (all builds are green)

### Resolved issues:
1. [x] resolves magento/magento2#29165: Add a head.additional block to adminhtml layout

### CODE REVIEW

**Comment by gsomoza** [comment]
Note1: this probably shouldn't have the `Area: Frontend` label.
Note2: assuming you're interested in this PR let me know if you'd like a backport for `2.3.x` (although as you can probably see it's quite trivial).

**Comment by ihor-sviziev** [comment]
@magento run all tests

**Review by ihor-sviziev — APPROVED** [review]

**Comment by magento-engcom-team** [comment]
Hi @ihor-sviziev, thank you for the review.
[ENGCOM-7718](https://jira.corp.magento.com/browse/ENGCOM-7718) has been created to process this Pull Request

**Comment by magento-engcom-team** [comment]
@gsomoza thank you for contributing. Please accept Community Contributors team invitation [here](https://github.com/orgs/magento/invitation) to gain extended permissions for this repository.

**Comment by gsomoza** [comment]
Done!

File: app/code/Magento/Backend/view/adminhtml/layout/default.xml
Comment (Nazar65) [inline app/code/Magento/Backend/view/adminhtml/layout/default.xml]: Maybe it's better to use `Magento\Backend\Block\Text\ListText` ? as this is adminhtml area ?

**Review by Nazar65 — CHANGES_REQUESTED** [review]

File: app/code/Magento/Backend/view/adminhtml/layout/default.xml
Comment (gsomoza) [reply app/code/Magento/Backend/view/adminhtml/layout/default.xml]: What's the added value of using that more-specific class over this one? I don't see that it adds much value because that child class doesn't actually "do" anything else other than extending the parent class, but if it gets deprecated (probably for that same reason) then it would be a BiC to go back to the parent version – while the opposite (going from the parent version to the child version) would be an easier change because it's not a BiC.

Comment (Nazar65) [reply app/code/Magento/Backend/view/adminhtml/layout/default.xml]: To be consistent with namespaces

Comment (Nazar65) [reply app/code/Magento/Backend/view/adminhtml/layout/default.xml]: and i don't see any reason for what this class can be deprecated, and this class defined in xml, so we can easily change to another one.

Comment (gsomoza) [reply app/code/Magento/Backend/view/adminhtml/layout/default.xml]: > this class defined in xml, so we can easily change to another one.

That's exactly my point: if we leave it as-is (using the framework type), then making changes in the future would be even easier than if we changed it to the backend type.

Being consistent with namespaces is a good principle when you're designing your type hierarchy. But it's less important when you're using it, especially if other principles like ISP could be more relevant.

The reasons why I'd leave it as-is:
1. ISP recommends using only the most meaningful contract, and nothing "more" than that.
2. Since we're using the more meaningful contract, which also happens to be the most generic type, it's also easier to make changes without causing BC breaks and there's more room for LSP (even though in practice it will never be applied at that level).
3. The framework type is the only type that carries actual meaning. The existence of the backend type that adds no value is questionable in the first place.

Having said that, I don't mind making the change if Magento really needs it to merge this code 🤷‍♂️. Are you in charge of making the final call on this PR or is this just a suggestion?

Comment (ihor-sviziev) [reply app/code/Magento/Backend/view/adminhtml/layout/default.xml]: @nazar65 in this case I do agree with @aredridel - we don’t really have any reason to use there class from backend module as it does the same. I would prefer keeping it as is

Comment (aredridel) [reply app/code/Magento/Backend/view/adminhtml/layout/default.xml]: ... Hm? With what?

Comment (ihor-sviziev) [reply app/code/Magento/Backend/view/adminhtml/layout/default.xml]: > ... Hm? With what?

With `Magento\Framework\View\Element\Text\ListText`

Comment (Nazar65) [reply app/code/Magento/Backend/view/adminhtml/layout/default.xml]: it was just suggestion, so if everyone  agree with `Magento\Framework\View\Element\Text\ListText` so just ignore my suggection

**Comment by engcom-Alfa** [comment]
@magento create issue

**Comment by engcom-Alfa** [comment]
:heavy_check_mark: QA Passed

**Manual testing scenario:**

1. Create a module that adds the following layout update in the **_adminhtml_** area:

```xml
<?xml version="1.0"?>
<page xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
      xsi:noNamespaceSchemaLocation="urn:magento:framework:View/Layout/etc/page_configuration.xsd">
    <referenceBlock name="head.additional">
        <block class="Magento\Framework\View\Element\Template" name="test"
               template="Magento_Backend::page/copyright.phtml"/>
    </referenceBlock>
</page>
```

2. Run: `php bin/magento cache:clean`;
3. Go to Admin and view page source;

**Before:** :heavy_multiplication_x: no copyright text in the `<head>`

Screenshot from 2020-07-16 11-19-05

**After:** :heavy_check_mark: The copyright text rendered within any admin page's <head> area

**Comment by naydav** [comment]
Moved to "on hold" status due to this PR requires some additional QA checks.
cc: @sdzhepa

**Comment by ihor-sviziev** [comment]
Could you explain which ones?

**Comment by naydav** [comment]
@ihor-sviziev 
We already have a few references to "head.additional" in scope of "adminhtml" area. 
Need to check how it's work without this fix?

Also, for frontend we use
https://github.com/magento-engcom/magento2/blob/6c7a031e6286ccd60d42a889280704be4ad1dce7/app/code/Magento/Catalog/view/frontend/layout/default.xml#L65
```
<block class="Magento\Framework\View\Element\Template" name="head.additional" as="head.additional" template="Magento_Theme::html/container.phtml"/>
```
> Because in the Frontend area this is done with a Template block with a container.phtml template is used, but that template doesn't exist in the adminhtml area.
Using of `ListText` (Adminhtml text list) is semantically incorrect.
Looks like more proper way is creating `container.phtml ` for admin area.

**Comment by ihor-sviziev** [comment]
Thx for explanation

**Comment by gsomoza** [comment]
> Using of ListText (Adminhtml text list) is semantically incorrect. Looks like a more proper way is creating some container.phtml for the admin area.

@naydav the purpose of `container.phtml` is to add a `<div>` tag around it's children. We don't want that inside `<head>` – I explained that in the description of the PR.

Now, if you want to add a `container.phtml` that does the exact same as `ListText` then let me know and I'll do it – but it seems like we're over-thinking this feature a little bit here 🤔

**Comment by ihor-sviziev** [comment]
@sidolov could you explain why this PR was moved to "pending review" without any comment? Can I approve it?

**Comment by sidolov** [comment]
@ihor-sviziev I don't see any reason to hold it, I agree with @naydav, additional cases should be verified but we can do that on the QA stage.

**Review by ihor-sviziev — APPROVED** [review]

**Comment by magento-engcom-team** [comment]
Hi @ihor-sviziev, thank you for the review.
[ENGCOM-7718](https://jira.corp.magento.com/browse/ENGCOM-7718) has been created to process this Pull Request

**Comment by engcom-Alfa** [comment]
Dev experience is required for testing this PR. Please note that Manual testing has not been performed.

**Comment by engcom-Foxtrot** [comment]
@magento run all tests

**Review by ihor-sviziev — APPROVED** [review]

**Comment by magento-engcom-team** [comment]
Hi @ihor-sviziev, thank you for the review.
[ENGCOM-7718](https://jira.corp.magento.com/browse/ENGCOM-7718) has been created to process this Pull Request

**Pull request merged**
